File: conference/management-bc/web/src/main/java/com/microsoft/conference/management/commandhandlers/ConferenceCommandHandler.cs
using System.Linq;
using System.Threading.Tasks;
using ConferenceManagement.Commands;
using ConferenceManagement.Domain.Models;
using ConferenceManagement.Domain.Services;
using ECommon.Components;
using ENode.Commanding;
using ENode.Infrastructure;

namespace ConferenceManagement.CommandHandlers
{
    [Component]
    public class ConferenceCommandHandler :
        ICommandHandler<CreateConference>,
        ICommandHandler<UpdateConference>,
        ICommandHandler<PublishConference>,
        ICommandHandler<UnpublishConference>,
        ICommandHandler<AddSeatType>,
        ICommandHandler<RemoveSeatType>,
        ICommandHandler<UpdateSeatType>,
        ICommandHandler<MakeSeatReservation>,
        ICommandHandler<CommitSeatReservation>,
        ICommandHandler<CancelSeatReservation>
    {
        private readonly ILockService _lockService;
        private readonly RegisterConferenceSlugService _registerConferenceSlugService;

        public ConferenceCommandHandler(ILockService lockService, RegisterConferenceSlugService registerConferenceSlugService)
        {
            _lockService = lockService;
            _registerConferenceSlugService = registerConferenceSlugService;
        }

        public Task HandleAsync(ICommandContext context, CreateConference command)
        {
            return Task.Factory.StartNew(() => _lockService.ExecuteInLock(typeof(ConferenceSlugIndex).Name, () =>
            {
                var conference = new Conference(command.AggregateRootId, new ConferenceInfo(
                    command.AccessCode,
                    new ConferenceOwner(command.OwnerName, command.OwnerEmail),
                    command.Slug,
                    command.Name,
                    command.Description,
                    command.Location,
                    command.Tagline,
                    command.TwitterSearch,
                    command.StartDate,
                    command.EndDate));
                _registerConferenceSlugService.RegisterSlug(command.Id, conference.Id, command.Slug);
                context.Add(conference);
            }));
        }
        public async Task HandleAsync(ICommandContext context, UpdateConference command)
        {
            var conference = await context.GetAsync<Conference>(command.AggregateRootId);
            conference.Update(new ConferenceEditableInfo(
                command.Name,
                command.Description,
                command.Location,
                command.Tagline,
                command.TwitterSearch,
                command.StartDate,
                command.EndDate));
        }
        public async Task HandleAsync(ICommandContext context, PublishConference command)
        {
            var conference = await context.GetAsync<Conference>(command.AggregateRootId);
            conference.Publish();
        }
        public async Task HandleAsync(ICommandContext context, UnpublishConference command)
        {
            var conference = await context.GetAsync<Conference>(command.AggregateRootId);
            conference.Unpublish();
        }
        public async Task HandleAsync(ICommandContext context, AddSeatType command)
        {
            var conference = await context.GetAsync<Conference>(command.AggregateRootId);
            conference.AddSeat(new SeatTypeInfo(
                command.Name,
                command.Description,
                command.Price), command.Quantity);
        }
        public async Task HandleAsync(ICommandContext context, RemoveSeatType command)
        {
            var conference = await context.GetAsync<Conference>(command.AggregateRootId);
            conference.RemoveSeat(command.SeatTypeId);
        }
        public async Task HandleAsync(ICommandContext context, UpdateSeatType command)
        {
            var conference = await context.GetAsync<Conference>(command.AggregateRootId);
            conference.UpdateSeat(
                command.SeatTypeId,
                new SeatTypeInfo(command.Name, command.Description, command.Price),
                command.Quantity);
        }
        public async Task HandleAsync(ICommandContext context, MakeSeatReservation command)
        {
            var conference = await context.GetAsync<Conference>(command.AggregateRootId);
            conference.MakeReservation(command.ReservationId, command.Seats.Select(x => new ReservationItem(x.SeatType, x.Quantity)).ToList());
        }
        public async Task HandleAsync(ICommandContext context, CommitSeatReservation command)
        {
            var conference = await context.GetAsync<Conference>(command.AggregateRootId);
            conference.CommitReservation(command.ReservationId);
        }
        public async Task HandleAsync(ICommandContext context, CancelSeatReservation command)
        {
            var conference = await context.GetAsync<Conference>(command.AggregateRootId);
            conference.CancelReservation(command.ReservationId);
        }
    }
}

File: samples/conference/management-bc/src/main/java/com/microsoft/conference/management/commandhandlers/ConferenceCommandHandler.java
package com.microsoft.conference.management.commandhandlers;

import com.microsoft.conference.common.management.commands.AddSeatType;
import com.microsoft.conference.common.management.commands.CancelSeatReservation;
import com.microsoft.conference.common.management.commands.CommitSeatReservation;
import com.microsoft.conference.common.management.commands.CreateConference;
import com.microsoft.conference.common.management.commands.MakeSeatReservation;
import com.microsoft.conference.common.management.commands.PublishConference;
import com.microsoft.conference.common.management.commands.RemoveSeatType;
import com.microsoft.conference.common.management.commands.UnpublishConference;
import com.microsoft.conference.common.management.commands.UpdateConference;
import com.microsoft.conference.common.management.commands.UpdateSeatType;
import com.microsoft.conference.management.domain.model.Conference;
import com.microsoft.conference.management.domain.model.ConferenceEditableInfo;
import com.microsoft.conference.management.domain.model.ConferenceInfo;
import com.microsoft.conference.management.domain.model.ConferenceOwner;
import com.microsoft.conference.management.domain.model.ReservationItem;
import com.microsoft.conference.management.domain.model.SeatTypeInfo;
import com.microsoft.conference.management.domain.service.RegisterConferenceSlugService;
import org.enodeframework.annotation.Command;
import org.enodeframework.annotation.Subscribe;
import org.enodeframework.commanding.ICommandContext;
import org.enodeframework.infrastructure.ILockService;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.stream.Collectors;

import static org.enodeframework.common.io.Task.await;

@Command
public class ConferenceCommandHandler {

    private ILockService lockService;

    @Autowired
    private RegisterConferenceSlugService registerConferenceSlugService;

    @Subscribe
    public void handleAsync(ICommandContext context, CreateConference command) {
        execInternal(context, command);
    }

    private void execInternal(ICommandContext context, CreateConference command) {
        Conference conference = new Conference(command.getAggregateRootId(), new ConferenceInfo(
                command.getAccessCode(),
                new ConferenceOwner(command.getOwnerName(), command.getOwnerEmail()),
                command.getSlug(),
                command.getName(),
                command.getDescription(),
                command.getLocation(),
                command.getTagline(),
                command.getTwitterSearch(),
                command.getStartDate(),
                command.getEndDate()));
        registerConferenceSlugService.registerSlug(command.getId(), conference.getId(), command.getSlug());
        context.add(conference);
    }

    @Subscribe
    public void handleAsync(ICommandContext context, UpdateConference command) {
        Conference conference = await(context.getAsync(command.getAggregateRootId(), Conference.class));
        conference.update(new ConferenceEditableInfo(
                command.name,
                command.description,
                command.location,
                command.tagline,
                command.twitterSearch,
                command.startDate,
                command.endDate));
    }

    @Subscribe
    public void handleAsync(ICommandContext context, PublishConference command) {
        Conference conference = await(context.getAsync(command.getAggregateRootId(), Conference.class));
        conference.publish();
    }

    @Subscribe
    public void handleAsync(ICommandContext context, UnpublishConference command) {
        Conference conference = await(context.getAsync(command.getAggregateRootId(), Conference.class));
        conference.unpublish();
    }

    @Subscribe
    public void handleAsync(ICommandContext context, AddSeatType command) {
        Conference conference = await(context.getAsync(command.getAggregateRootId(), Conference.class));
        conference.addSeat(new SeatTypeInfo(
                command.name,
                command.description,
                command.price), command.quantity);
    }

    @Subscribe
    public void handleAsync(ICommandContext context, RemoveSeatType command) {
        Conference conference = await(context.getAsync(command.getAggregateRootId(), Conference.class));
        conference.removeSeat(command.getSeatTypeId());
    }

    @Subscribe
    public void handleAsync(ICommandContext context, UpdateSeatType command) {
        Conference conference = await(context.getAsync(command.getAggregateRootId(), Conference.class));
        conference.updateSeat(
                command.seatTypeId,
                new SeatTypeInfo(command.name, command.description, command.price),
                command.quantity);
    }

    @Subscribe
    public void handleAsync(ICommandContext context, MakeSeatReservation command) {
        Conference conference = await(context.getAsync(command.getAggregateRootId(), Conference.class));
        conference.makeReservation(command.reservationId, command.seats.stream().map(x -> new ReservationItem(x.seatType, x.quantity)).collect(Collectors.toList()));
    }

    @Subscribe
    public void handleAsync(ICommandContext context, CommitSeatReservation command) {
        Conference conference = await(context.getAsync(command.getAggregateRootId(), Conference.class));
        conference.commitReservation(command.reservationId);
    }

    @Subscribe
    public void handleAsync(ICommandContext context, CancelSeatReservation command) {
        Conference conference = await(context.getAsync(command.getAggregateRootId(), Conference.class));
        conference.cancelReservation(command.reservationId);
    }
}
